src/vcard/mod_vcard_api.erl

```diff
@@ -0,0 +1,277 @@
+
+-include("mongoose.hrl").
+-include("jlib.hrl").
+-include("mod_vcard.hrl").
+
+-export([transform_from_graphql/1,
+         set_vcard/4,
+         to_gql_format/1,
+         get_vcard/3,
+         name_components_proccess/2,
+         address_components_proccess/2,
+         label_components_proccess/2,
+         telephone_components_proccess/2,
+         email_components_proccess/2,
+         geo_components_proccess/2,
+         org_components_proccess/2
+        ]).
+
+set_vcard(HostType, LUser, LServer, Vcard) ->
+    mod_vcard:unsafe_set_vcard(HostType, #jid{user = LUser, lserver = LServer}, transform_from_graphql(Vcard)).
+
+get_vcard(HostType, LUser, LServer) ->
+    try mod_vcard_backend:get_vcard(HostType, LUser, LServer) of
+        {ok, VCARD} ->
+               [#xmlel{children = VcardData}] = VCARD,
+               to_gql_format(VcardData);
+        {error, Reason} ->
+               {error, Reason}
+        catch E:R:Stack ->
+            ?LOG_ERROR(#{what => vcard_sm_iq_get_failed,
+                         class => E, reason => R, stacktrace => Stack}),
+            {error, error}
+        end.
+
+transform_from_graphql(Vcard) ->
+    #xmlel{name = <<"vCard">>,
+           attrs = [{<<"xmlns">>, <<"vcard-temp">>}],
+           children = lists:foldl(fun({Name, Value}, Acc) ->
+                                      Acc ++ transform_field_and_value(Name, Value)
+                                  end, [], maps:to_list(Vcard))}.
+
+construct_xmlel(Name, Children) when is_list(Children)->
+    [#xmlel{name = Name,
+            attrs = [],
+            children = Children}].
+
+transform_field_and_value(_Name, null) ->
+    [];
+transform_field_and_value(Name, Value) when is_list(Value) ->
+    lists:foldl(fun(Element, Acc) ->
+                    Acc ++ transform_field_and_value(Name, Element)
+                end, [], Value);
+transform_field_and_value(Name, Value) when is_map(Value) ->
+    construct_xmlel(from_gql_to_xml(Name), proccess_child_map(Value));
+transform_field_and_value(Name, Value) ->
+    construct_xmlel(from_gql_to_xml(Name), [{xmlcdata, Value}]).
+
+transform_subfield_and_value(_Name, null) ->
+    [];
+transform_subfield_and_value(<<"tags">>, TagsList) ->
+    lists:foldl(fun(Tag, Acc) ->
+                    Acc ++ construct_xmlel(Tag, [])
+                end, [], TagsList);
+transform_subfield_and_value(Name, Value) when is_list(Value) ->
+    lists:foldl(fun(Element, Acc) ->
+                    Acc ++ construct_xmlel(from_gql_to_xml(Name), [{xmlcdata, Element}])
+                end, [], Value);
+transform_subfield_and_value(Name, Value) ->
+    construct_xmlel(from_gql_to_xml(Name), [{xmlcdata, Value}]).
+
+proccess_child_map(Value) ->
+    io:format("~p", [maps:to_list(Value)]),
+    lists:foldl(fun({Name, SubfieldValue}, Acc) ->
+                    Acc ++ transform_subfield_and_value(Name, SubfieldValue)
+                end, [], maps:to_list(Value)).
+
+from_gql_to_xml(<<"formattedName">>) -> <<"FN">>;
```

Comment:
Yeah, the only name conversion looks better than it was done before for the whole XML :D

I am wondering if we could have one clause for the names that only need the capitalization.